### PR TITLE
(PXP-5787) Support Shepherd API

### PR DESCRIPTION
**Pull request by mpingram**
Ticket: https://ctds-planx.atlassian.net/browse/PXP-5787

Updates `gen3-client` to interact with the [Shepherd API](https://docs.google.com/document/d/1vRrM5xvTepzRqw3tQpqR8hLg4QdbP3eD44_bKZsSL04/edit#) if the Shepherd API is deployed to the commons.

**Testing these changes against the Shepherd API**
My dev environment (https://mpingram.planx-pla.net) has a basic version of the Shepherd API deployed (uploading files with `POST /objects` is implemented, but downloading files is not yet implemented.) To test these changes against the Shepherd API, you can:
- clone this branch
- create a test file (`testfile.txt`) and a test metadata file (`testfile_metadata.json`, use `/programs/DEV/projects/test` for the authz)
- create a profile with Shepherd support enabled:
```
 gen3-client configure --profile=testprofile --cred=/PATH/TO/CRED --apiendpoint=https://mpingram.planx-pla.net --use-shepherd=true --min-shepherd-version=1.0.0
```
- Run this command:
```
go run main.go upload --profile=testprofile --upload-path=./testfile.txt --metadata
```

**Future work required:**
- This PR doesn't change the multipart upload code to work with Shepherd, as the Shepherd API doesn't yet include support for multipart uploads.
- Need to improve the UX of uploading metadata alongside files once the file upload flow with the Shepherd API becomes more clear. Currently, the user is required to manually add a `[filename]_metadata.json` file alongside each file with the metadata. We can extend the gen3-client with more tooling to make this easier (e.g. a `generate-metadata` command), but I think it makes sense to do that later in a separate PR: We're not sure what the file upload flow will look like in an all-Shepherd commons yet. (Plus this PR is already long 😅)

### New Features
- The gen3-client `upload`, `download-single`, and `download-multiple` commands now interact with the Shepherd API if it is deployed.
- The gen3-client `upload` command can be used to upload file metadata (`authz`, `aliases`, and arbitrary metadata for `metadataservice`) if uploading to the Shepherd API. See the [README](https://github.com/uc-cdis/cdis-data-client/blob/feat/pxp-5787-support-shepherd-api/README.md#uploading-file-metadata) for documentation.

### Improvements
- Added unit tests for some internal functions that interact with Gen3 APIs

### CODE REVIEW

File: gen3-client/commonUtils/commonUtils.go [hunk 1]
```diff
@@ -15,6 +15,15 @@ import (
...
 )
 
+// ShepherdEndpoint is the endpoint postfix for SHEPHERD / the Object Management API
+const ShepherdEndpoint = "/mds"
+
+// ShepherdVersionEndpoint is the endpoint used to check what version of Shepherd a commons has deployed
+const ShepherdVersionEndpoint = "/mds/version"
+
+// MinAcceptableShepherdVersion is the minimum version of Shepherd that will work with this version of the client.
+const MinAcceptableShepherdVersion = "2.0.0"
```
Comment (mfshao) [inline gen3-client/commonUtils/commonUtils.go:25]: maybe should make this configurable, modify the `configure` cmd to take an optional input from user and save it with the profile

File: gen3-client/g3cmd/upload.go [hunk 2]
```diff
@@ -134,9 +152,10 @@ func init() {
 			if len(multipartFilePaths) > 0 {
 				log.Println("Multipart uploading....")
 				for _, filePath := range multipartFilePaths {
-					fileInfo, err := ProcessFilename(uploadPath, filePath, includeSubDirName)
+					// NOTE @mpingram -- metadata upload will not work for multipart uploads until Shepherd API supports multipart uploads
+					fileInfo, err := ProcessFilename(uploadPath, filePath, includeSubDirName, false)
```
Comment (mfshao) [inline gen3-client/g3cmd/upload.go:156]: i'd prefer to also check if we are using shepherd here and print some msgs && exiting if that is the case

File: gen3-client/g3cmd/utils.go [hunk 3]
```diff
@@ -599,3 +698,28 @@ func FormatSize(size int64) string {
...
+// Gen3Interface contains methods used to make authorized http requests to Gen3 services.
+type Gen3Interface interface {
+	CheckForShepherdAPI(profile string) (bool, error)
+	GetResponse(profile string, configFileType string, endpointPostPrefix string, method string, contentType string, bodyBytes []byte) (string, *http.Response, error)
+	DoRequestWithSignedHeader(profile string, configFileType string, endpointPostPrefix string, contentType string, bodyBytes []byte) (jwt.JsonMessage, error)
+	MakeARequest(method string, apiEndpoint string, accessKey string, contentType string, headers map[string]string, body *bytes.Buffer) (*http.Response, error)
+}
+
+// NewGen3Interface returns a struct that contains methods used to make authorized http requests to Gen3 services.
+func NewGen3Interface() Gen3Interface {
```
Comment (mfshao) [inline gen3-client/g3cmd/utils.go:711]: this is awesome! definately better than passing `function` around

File: gen3-client/g3cmd/utils.go [hunk 4]
```diff
@@ -490,27 +586,30 @@ func initBatchUploadChannels(numParallel int, inputSliceLen int) (int, chan *htt
 }
 
 func batchUpload(furObjects []commonUtils.FileUploadRequestObject, workers int, respCh chan *http.Response, errCh chan error) {
+	// Instantiate interface to Gen3
+	gen3Interface := NewGen3Interface()
```
Comment (mfshao) [inline gen3-client/g3cmd/utils.go:590]: i think for `batchDownload` you passed ref of a `Gen3Interface` into that function to reuse, should we do the same for `batchUpload`?

File: gen3-client/jwt/functions.go [hunk 5]
```diff
@@ -132,6 +139,37 @@ func (f *Functions) ParseFenceURLResponse(resp *http.Response) (JsonMessage, err
...
+		return false, fmt.Errorf("Error occurred when parsing version from Shepherd: %v: %v", string(body), err)
+	}
+	minVer, err := semver.Parse(commonUtils.MinAcceptableShepherdVersion)
+	if err != nil {
+		return false, fmt.Errorf("Error occurred when parsing minimum acceptable Shepherd version: %v: %v", commonUtils.MinAcceptableShepherdVersion, err)
+	}
+	if ver.GTE(minVer) {
+		return true, nil
+	}
+	return false, nil
```
Comment (mfshao) [inline gen3-client/jwt/functions.go:170]: should we return an error in this case? like `version too low blabla...`

File: gen3-client/g3cmd/configure.go [hunk 6]
```diff
@@ -66,5 +79,7 @@ func init() {
 	configureCmd.MarkFlagRequired("cred")
 	configureCmd.Flags().StringVar(&apiEndpoint, "apiendpoint", "", "Specify the API endpoint of the data commons")
 	configureCmd.MarkFlagRequired("apiendpoint")
+	configureCmd.Flags().StringVar(&useShepherd, "use-shepherd", "", fmt.Sprintf("Enables or disables support for the Shepherd API. If enabled, gen3client will use the Shepherd API if available. (Default: %v)", commonUtils.DefaultUseShepherd))
+	configureCmd.Flags().StringVar(&minShepherdVersion, "min-shepherd-version", "", fmt.Sprintf("Specify the minimum version of Shepherd that the gen3client will use if Shepherd is enabled. (Default: %v)", commonUtils.DefaultMinShepherdVersion))
```
Comment (mfshao) [inline gen3-client/g3cmd/configure.go:83]: put this into docs as well?

File: go.mod [hunk 7]
```diff
@@ -3,6 +3,7 @@ module github.com/uc-cdis/gen3-client
 go 1.14
 
 require (
+	github.com/blang/semver/v4 v4.0.0
```
Comment (mfshao) [inline go.mod:6]: just curious have you compared this with https://github.com/hashicorp/go-version ?

https://github.com/hashicorp/go-version is a dep we are already importing so not sure if we can just use that? But this is not a big deal

**Review by mfshao — COMMENTED** [review]
thank you! it generally lgtm, just 2 minor comments

File: gen3-client/g3cmd/configure.go (hunk 6 above)
Comment (mpingram) [reply gen3-client/g3cmd/configure.go:83]: I put a section in the README, anything else I should add? https://github.com/uc-cdis/cdis-data-client/pull/73/files#diff-04c6e90faac2675aa89e2176d2eec7d8R41-R47

File: go.mod (hunk 7 above)
Comment (mpingram) [reply go.mod:6]: 🤦 I hadn't realized we were already importing `go-version`! The package APIs look almost identical, I'll try to swap the dependency out for `go-version.`

File: gen3-client/g3cmd/configure.go (hunk 6 above)
Comment (mfshao) [reply gen3-client/g3cmd/configure.go:83]: oh it is just the README seems don't have anything about the `min-shepherd-version` arg 😅

Comment (mpingram) [reply gen3-client/g3cmd/configure.go:83]: Oh gotchu, yes I'll add that!

**Review by mfshao — APPROVED** [review]
cool! 👍

**Pull request merged**